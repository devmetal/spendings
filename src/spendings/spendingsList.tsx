import React, { FC } from 'react';
import { TSpending } from '../types';
import { Spending } from './spending';

export interface SpendingsListProps {
  /**
   * List of user spendings
   */
  spendings: Array<TSpending>;
  /**
   * User start update
   */
  onUpdate: (spending: TSpending) => void;
  /**
   * User start delete
   */
  onDelete: (id: number) => void;
}

/**
 * Spendings is a basic list of represent user spendings
 */
export const SpendingsList: FC<SpendingsListProps> = ({
  spendings,
  onUpdate,
  onDelete,
}) => {
  /*const yPos = useRef(0);
  const mOffset = useRef(0);
  const container = useRef<HTMLDivElement | null>(null);

  useLayoutEffect(() => {
    if (container !== null && container.current !== null) {
      mOffset.current = -1 * container.current.offsetHeight;
    }
  }, []);

  const [props, set] = useSpring(() => ({ y: 0 }));

  const bind = useDrag(
    ({ down, delta: [, yDelta], cancel = () => {} }) => {
      yPos.current += yDelta;

      if (yPos.current > 0 && !down) {
        cancel();
        yPos.current = 0;
      }

      if (yPos.current < mOffset.current && !down) {
        cancel();
        yPos.current = mOffset.current;
      }

      set({ y: yPos.current });
    },
    { lockDirection: true, axis: 'y' }
  );*/

  return (
    <div>
      {spendings.map((spending) => (
        <Spending
          key={spending.id}
          id={spending.id}
          name={spending.name}
          amount={spending.amount}
          onUpdate={onUpdate}
          onDelete={onDelete}
        />
      ))}
    </div>
  );
};
