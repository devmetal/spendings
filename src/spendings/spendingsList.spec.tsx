import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { SpendingsList } from './spendingsList';

const delay = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

const props = {
  onDelete: jest.fn(),
  onUpdate: jest.fn(),
  spendings: [
    {
      id: '0',
      amount: 10,
      name: 'test',
    },
  ],
};

test('Sanity snapshot testing', () => {
  const { container } = render(<SpendingsList {...props} />);
  expect(container).toMatchSnapshot();
});

describe('interactions', () => {
  let updateBtn: HTMLElement;
  let deleteBtn: HTMLElement;

  beforeEach(() => {
    const { getByRole } = render(<SpendingsList {...props} />);
    updateBtn = getByRole('update-item');
    deleteBtn = getByRole('delete-item');
  });

  test('update interaction show update ui', async () => {
    fireEvent.click(updateBtn);
    await waitFor(() => expect(screen.getByText('Cancel')).toBeVisible());
  });

  test('hold delete button more than 1500ms should call onDelete', async () => {
    const evt = document.createEvent('MouseEvents');
    evt.initEvent('mousedown', true, true);
    fireEvent(deleteBtn, evt);
    await delay(1800);
    expect(props.onDelete).toBeCalled();
  });
});
