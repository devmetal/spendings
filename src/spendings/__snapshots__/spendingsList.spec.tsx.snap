// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sanity snapshot testing 1`] = `
<div>
  <div>
    <div
      class="makeStyles-root-1"
    >
      <div
        class="MuiPaper-root MuiCard-root makeStyles-card-2 MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="MuiCardContent-root"
        >
          <p
            class="MuiTypography-root makeStyles-name-3 MuiTypography-body1"
          >
            test
          </p>
          <p
            class="MuiTypography-root makeStyles-amount-4 MuiTypography-body1"
          >
            10
             HUF
          </p>
        </div>
        <div
          class="MuiCardActions-root MuiCardActions-spacing"
        >
          <button
            aria-label="update"
            class="MuiButtonBase-root MuiFab-root MuiFab-sizeSmall MuiFab-primary"
            role="update-item"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiFab-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <div
            class="makeStyles-wrapper-5"
          >
            <button
              aria-label="delete"
              class="MuiButtonBase-root MuiFab-root MuiFab-sizeSmall MuiFab-secondary"
              role="delete-item"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiFab-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <div
              aria-valuenow="0"
              class="MuiCircularProgress-root makeStyles-fabProgress-6 MuiCircularProgress-colorPrimary MuiCircularProgress-determinate"
              role="progressbar"
              style="width: 50px; height: 50px; transform: rotate(-90deg);"
            >
              <svg
                class="MuiCircularProgress-svg"
                viewBox="22 22 44 44"
              >
                <circle
                  class="MuiCircularProgress-circle MuiCircularProgress-circleDeterminate"
                  cx="44"
                  cy="44"
                  fill="none"
                  r="20.2"
                  stroke-width="3.6"
                  style="stroke-dasharray: 126.920; stroke-dashoffset: 126.920px;"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
